### ISSUE TITLE
Use Protobuf and gRPC for internal communications

### ISSUE THREAD

**Issue by yurishkuro** [jaegertracing/jaeger#773]
## Problem

At the moment we have the following data models handled by the backend, often with two-way transformers between them:
  * Domain model in `/model` used by all internal APIs and business logic
    * JSON serialization of the domain model is used in many unit and integration tests
  * `jaeger.thrift` as the main on-the-wire representation
  * `/model/json` is JSON used for communicating with the UI
  * `/model/json` with embedded Process used by ElasticSearch storage plugin
  * `zipkin.thrift` and Zipkin JSON accepted as input data models as well

We also have limitations on routing traffic between agents and collectors because they use TChannel as the communication protocol.

### Converge on a Protobuf 3 Model

By converging on a single Protobuf model we can consolidate some of these models, reduce the number of converters, and unblock / enable several other work streams:
  * The JSON format exposed by the query service can be based on standard [Protobuf JSON Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json) as Jaeger's canonical JSON model (issue https://github.com/jaegertracing/jaeger/issues/706).
  * The collector can accept the same JSON format, which will unblock the work on pure Javascript client (issue https://github.com/jaegertracing/jaeger-client-javascript/issues/1)
  * The Protobuf model can be used to generate official documentation for the API, for gRPC and REST endpoints (issue https://github.com/jaegertracing/jaeger/issues/456).
  * The domain model can be replaced with code generated from Protobuf IDL (already done in https://github.com/jaegertracing/jaeger/issues/856). The canonical JSON representation can be used for unit / integration tests (also done).
  * The UI can be upgraded to use canonical JSON model, eliminating the need for `/model/json` (the ElasticSearch plugin will need to retain the model and converters, which should be migrated into the es module)

### Use gRPC with Protobuf model

* Since gRPC is based on HTTP/2, it has much wider support than TChannel, e.g. for routing via proxies
* gRPC can support TLS, addressing issues https://github.com/jaegertracing/jaeger/issues/458 and https://github.com/jaegertracing/jaeger/issues/404
* gRPC can be used to to serve both binary and REST endpoints, and generate documentation and Swagger API

### Phase 1 - MVP

Goal: deprecate TChannel.

* [ ] Support gRPC/protobuf endpoints for core backend communications
  * [x] Collector service (receiving spans)
  * [ ] Query service
  * [x] Configuration services in the collector (sampling, throttling, baggage restrictions)
* [x] All agent to collector communications using gRPC
  * [x] Spans forwarding
  * [x] Configuration services
* [ ] Lock down protobuf-based API 2.x

### Phase 2

* [ ] UI uses gRPC-based REST endpoint instead of current bespoke JSON
* [ ] Publish Swagger (nice to have)

### Phase 3

* [ ] Agent exposes gRPC endpoints, to enable clients work

## Implementation notes

We've talked about this a lot, but there are many moving pieces that must come together, so this is the meta issue to list them all out, in random order.

* We want to have a single protobuf data model on the server
* We want that data model to be representable in JSON so that query service API can be expressed with it instead of the current JSON model (issue #706)
* We probably want to use gogoprotobuf instead of the default protobuf because gogo allows generation of the data model without too many pointers, e.g. `[]model.KeyValue` instead of `[]*model.KeyValue` ([gist][gist])
* We need to decide if we want to use `oneof` for KeyValue or keep sending the Type field explicitly as we do today
* We need to decide if we want to keep the existing UI model's optimization where the Process object in the Span is deduped and replaced with a [reference][1] to an entry in the map of Process objects attached to the Trace
  * The current UI model supports both embedded Process and ProcessID reference, so we could keep this dual representation
  * The deduping of the Process objects is more important when reporting spans from clients to the backend (which uses the [Batch type][2]) than it is for the query service API. The UI does not really rely on the deduping, it's primarily done for the purpose of reducing the JSON message size when returning the whole trace.
* My proposal is to keep both Batch and Trace types. Batch doesn't need any changes.
* Need to investigate how to expose JSON API for the UI as a facade on top of a gRPC API implemented by the query service
  * https://github.com/grpc-ecosystem/grpc-gateway seems to work, apparently can even generate swagger
  * https://github.com/yarpc/yarpc-go could be another possibility, as it also provides client side load balancing and integration with discovery. We have that in the current tchannel-based implementation, but gRPC doesn't provide it out of the box (or does it?)
* The data model generated with gogoprotobuf is actually very close to the existing domain model `model.Span`. We need to decide if we want to keep them separate and do transformation or merge them into one by generating the core skeletons and then adding the additional business methods via companion hand-written source files. I am in favor of merging them to minimize data copying. The risk here is if future changes in proto/gogo would result in differently generated model, but that seems unlikely.
* Current comms between agent and collectors are done with tchannel that supports client side load balancing and integration with discovery. We'd need an equivalent of that with gRPC, e.g. by using yarpc.
* There are a few differences in the JSON generated from IDL in the [gist][gist] (some could be addressed via custom renderers):
  * dates are rendered in UTC: "2018-04-02T02:17:45.171667100Z"
  * durations are rendered as string: "0.013s"
  * trace ID encoded as two integer fields instead of one string (some int64s cannot be parsed by JS)
  * span ID encoded as integer instead of one string (some int64s cannot be parsed by JS)
  * `KeyValue.VType=STRING` and `SpanRef.Type=CHILD_OF` are omitted by default (as zero values)

## Implementation plan

* [x] define official proto #856
* [x] merge domain model with proto-generated model #856
  * fix tests to use proto-generated JSON instead of the current `encoding/json`
* [ ] Define project layout for Protobuf / gRPC / generated files #904
* [ ] redefine collector API via proto IDL and implement it with gRPC
* [ ] Combine gRPC and HTTP handlers on the same port
  * https://github.com/philips/grpc-gateway-example/blob/master/cmd/serve.go
  * https://coreos.com/blog/gRPC-protobufs-swagger.html
* [ ] add new reporter to the agent using gRPC/proto model
* [ ] redefine query service API via proto IDL and implement it with gRPC
* [ ] Serve Swagger page
* [ ] implement a REST/JSON facade on top of gRPC
* [ ] upgrade UI to the new JSON data model
* [ ] implement other collector APIs with gRPC
  * sampling, throttling, etc.
* [ ] implement agent APIs with gRPC
* [ ] implement gRPC transport in the clients

[1]: https://github.com/jaegertracing/jaeger/blob/master/model/json/model.go#L79
[2]: https://github.com/jaegertracing/jaeger/blob/master/thrift-gen/jaeger/ttypes.go#L1582
[gist]: https://gist.github.com/yurishkuro/d2220a386a82c91f07c13bcb664f4165

**Comment by yurishkuro** [comment]
For Jaeger UI we might be able to use https://github.com/grpc/grpc-web. But I still think we need native JSON support, if only for jaeger-client-javascript to be able to report traces from the browser. I don't think it would be a good idea to insist on having grpc-web in the browser applications as a dependency, given that JSON is a lot more natural and easily available (possibly for other languages as well).

**Comment by jpkrohling** [comment]
There's gRPC Gateway, which can take care of providing transparent REST+JSON support: https://github.com/grpc-ecosystem/grpc-gateway

**Comment by tiffon** [comment]
Re jaeger-client-javascript, using protobufs, directly, is also an option:

https://github.com/dcodeIO/protobuf.js

Has broad browser compatibility: http://dcode.io/protobuf.js/#compatibility

**Comment by isaachier** [comment]
So I briefly looked into this last night and the consensus was that gzipped JSON will almost always beat Javascript protobuf.

**Comment by tiffon** [comment]
>consensus was that gzipped JSON will almost always beat Javascript protobuf.

@isaachier Can you refer to relevant material? And, beat in what sense, speed, size, etc?

**Comment by isaachier** [comment]
OK consensus might be an overstatement. See this article: https://blog.wearewizards.io/using-protobuf-instead-of-json-to-communicate-with-a-frontend.

**Comment by isaachier** [comment]
For the record, I wouldn't use node as a good indicator of frontend performance. Node C++ extensions can easily add speed ups that aren't available in the browser.

**Comment by yurishkuro** [comment]
I believe we need to support JSON format for the following reasons:
  * Submitting JSON from clients (e.g. from browser) usually requires fewer (if any) dependencies than protobuf
  * ElasticSearch storage uses JSON as the data model on the wire
  * UI developer experience is a lot better (I assume) when the server exposes JSON API than with protobuf, e.g. you can see the payloads in browser dev tools
  * External analytical tools would probably have easier time consuming JSON than protobuf
  * Even our internal unit tests represent traces in JSON, which is readable & easier to author than protobuf

**Comment by yurishkuro** [comment]
Regarding Go rendering of duration in JSON (e.g. `1666h40m0.016000013s`) - another alternative is to record end timestamp instead of duration. 
Pros:
  * no format issues
  * semantically works better when we support partial spans
Cons:
  * more data on the wire: +4 bytes in protobuf, and full UTC timestamp in JSON.

**Comment by isaachier** [comment]
I wouldn't worry too much about the size of the span on the wire. As long as it doesn't hurt performance somewhere, people generally don't care.

**Comment by isaachier** [comment]
Fun fact:

> Proto3 supports a canonical encoding in JSON, making it easier to share data between systems. The encoding is described on a type-by-type basis in the table below.
https://developers.google.com/protocol-buffers/docs/proto3#json

**Comment by tiffon** [comment]
For duration, why not just use nanoseconds?

**Comment by yurishkuro** [comment]
Nanoseconds are fine in 64bit, but Javascript only understands 53bit and that's 100 days in nanos - uncomfortably low number. The trade-off is we either render 64bit duration to JSON as a string, e.g. `1666h40m0.016000013s`, or we replace duration with `end_timestamp`.

**Comment by yurishkuro** [comment]
I suppose another option would be to render int64 as a string in JSON, to avoid precision loss in Javascript.

**Comment by tiffon** [comment]
If this is our canonical storage model, too, that would make it more difficult to do duration based queries. 

What about storing the duration as milliseconds as a float?

**Comment by tiffon** [comment]
FYI, milliseconds as a float is the approach taken by the [`Performance`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) API.

**Comment by isaachier** [comment]
I'm wondering if we'd care about the precision for that. Otherwise, [`struct timespec`](http://en.cppreference.com/w/c/chrono/timespec) is probably the underlying implementation for all of these languages so we can copy it to Protobuf.

**Comment by yurishkuro** [comment]
Fwiw, average span size in our deployment is between 400-500 bytes (in thrift), so we'd be adding 1% overhead on the data size by switching from duration to timestamp.

**Comment by simonpasquier** [comment]
My 2 cents from Prometheus experience.

> We probably want to use gogoprotobuf instead of the default protobuf because gogo allows generation of the data model without too many pointers, e.g. []model.KeyValue instead of []*model.KeyValue (gist) 

Prometheus uses gogoprotobuf with no issue.

> https://github.com/grpc-ecosystem/grpc-gateway seems to work, apparently can even generate swagger

Prometheus also uses `grpc-gateway` for its v2 API. The API is experimental and provided without stability guarantee but it is working fine and indeed there's even a [Swagger documentation](https://github.com/prometheus/prometheus/blob/master/documentation/dev/api/swagger.json) generated from the Protobuf definition. More details in this [Google doc](https://docs.google.com/document/d/11j6gBolmCftTo36COdCVl9qztx5juOIz8Fuo2nx_Qvw/edit).

**Comment by tiffon** [comment]
Seems the main benefits of using floats to store the duration in milliseconds is simplicity; ease to support duration based queries; it's a bit more aligned, semantically, with the property we're interested in storing and string parsing is not required in any context.

**Comment by johanbrandhorst** [comment]
> KeyValue.VType=STRING and SpanRef.Type=CHILD_OF are omitted by default (as zero values)

This is configurable in the [marshaler](https://github.com/gogo/protobuf/blob/1ef32a8b9fc3f8ec940126907cedb5998f6318e4/jsonpb/jsonpb.go#L65). I would recommend using `OrigName` as well for snake_case variable names.

Note that committing to a JSON representation limits your options for backwards compatible changes (no field renaming). This is a subtle difference from usual proto3 forward/backward compatibility benefits.

**Comment by johanbrandhorst** [comment]
@yurishkuro note that if the JSON representation is _very_ important to be in a specific format, you can use the `google.protobuf.Struct` from the [`struct.proto`](https://github.com/google/protobuf/blob/master/src/google/protobuf/struct.proto) _Well Known Type_. These are meant to help with JSON schema design. It obviously compromises your proto types a little, but it can be done.

**Comment by johanbrandhorst** [comment]
If adding a new WellKnownType to your protofile, you'll need to adjust the `protoc` generation steps; you'll need another `-Mgoogle/protobuf/struct.proto:github.com/gogo/protobuf/types` to each `x_out` invocation.

**Comment by johanbrandhorst** [comment]
I would at the same time consider replacing the `Makefile` with [prototool](https://github.com/uber/prototool) :smile:.

**Comment by peter-edge** [comment]
I recommend against using grpc-gateway for this purpose. Contact me offline for alternatives.

**Comment by johanbrandhorst** [comment]
@peter-edge mind sharing your rationale? I think the grpc-gateway is an increasingly popular choice for this kind of thing and would love any insights you might have.

**Comment by yurishkuro** [comment]
@jaegertracing/jaeger-maintainers I've reached an indecision point with #856. I am fairly happy with gogo-generated classes from the proto definition and can use them as a replacement for the domain model, which is great as it removes one (almost always present) conversion step. But I am a bit stuck with JSON serialization. Currently I have custom marshalers that serialize trace-id and span-id as hex strings (even though internally trace-id is a struct of two uint64's and span-id is uint64). I had one remaining todo to figure out serialization of the duration field. But what I don't like is that while I can do all those custom serializations in Go, they are not as easy, if at all possible, to reproduce in other languages. It would be much simpler to just let protobuf do its default JSON serialization ([proto3-json]), which should be compatible across languages. It would also have better compatibility long term with tools like grpc-gateway, swagger, etc. because the JSON would match the proto definition (currently trace-id doesn't). Specifically, these would be the differences from the current jaeger ui JSON:
  * trace-id will be a base64 string instead of hex string. I would redefine it in proto as `bytes`.
  * potentially the same with the span-id
  * duration is a string like "1.000340012s", "1s"

The base64 encoding of trace/span-id is actually the most problematic because everywhere else we're using hex strings.

[proto3-json]: https://developers.google.com/protocol-buffers/docs/proto3#json

**Comment by peter-edge** [comment]
I highly recommend using the protobuf JSON mapping - not only is it repeatable cross-language with each protobuf implementation, there’s actually some edge cases where correctness is an issue.

**Comment by isaachier** [comment]
Isn't the JSON only significant for UI? If Javascript is the only problematic language, then we would just need to find a solution to that. Any custom encoding/decoding sounds like something worth avoiding.

**Comment by yurishkuro** [comment]
@isaachier we are trying to define the official, long term API, which in theory could be consumed by other tools, not just the Jaeger UI. But at the moment it is just the UI, and in fact in #856 I haven't changed its custom representation that's different from the domain model.

BTW, agreeing to use `bytes` for IDs with base64 encoding makes the model similar to zipkin and census proto models. It's also more easily extensible for longer IDs if needed. I can still use a custom type like `type TraceID [16]byte` to avoid memory allocations in the domain model.

**Comment by Falco20019** [comment]
The problem with default deserialization of durations and timestamps in JSON is that some languages use `ms` and some use `ns` as unit... I often had problem between `Java` (serializing `ns`) and `C#` (serializing `ms`) and also made an PR for `RestSharp` a year ago to accept both. Guessing works somewhat fine for timestamps but not so good for durations. The string is less error prone in that specific case.

**Comment by yurishkuro** [comment]
> That's also for performance, since in the proto-wire format, the default values are omitted to improve performance.

I actually liked that aspect that both CHILD_OF and STRING are not included in the JSON since they occur in 99% of cases and are very good defaults.

**Comment by isaachier** [comment]
Regarding snake case: https://developers.google.com/protocol-buffers/docs/style.

**Comment by Falco20019** [comment]
> I actually liked that aspect that both CHILD_OF and STRING are not included in the JSON since they occur in 99% of cases and are very good defaults.

Ok, the initial post sounded like you want to get rid of that behaviour (stating `some could be addressed via custom renderers`). Because I also like avoiding useless data transfers.

@isaachier Thanks for the link to the style guide. Totally forgot that it's also explicitly documented :)

**Comment by peter-edge** [comment]
You'll probably want to follow the Uber Style Guide, enforced by Prototool :-) https://github.com/uber/prototool/blob/dev/etc/style/uber/uber.proto you can think of it as "strict mode" as compared to the above.

**Comment by yurishkuro** [comment]
For reference:  Docker images for protoc compiler https://github.com/namely/docker-protoc

**Comment by pavolloffay** [comment]
I am working on this now. I want to implement a minimal working solution using grpc communication between agent and collector and get it merged to master as experimental. 

So far I have rebased @isaachier PR https://github.com/jaegertracing/jaeger/pull/1042 and added proto for sampling. I will be submitting PR shortly.

I will be posting my notes here what should be done afterwards:

- [ ] metric:"http-server.errors" tags:"status=5xx,source=tcollector-proxy
- [] refactor sampling store to use proto model or internal model only for the store

**Comment by yurishkuro** [comment]
> added proto for sampling

Would it move this along better if we start merging smaller pieces into master, e.g. my branch & Isaac's PR? I think the tests were outstanding.

**Comment by pavolloffay** [comment]
I think it would break me right now. I will submit a PR to master based on those two. Now fixing the coverage.

**Comment by isaachier** [comment]
Hey @pavolloffay. Glad to see this finally happening! Please let me know if you need any help here.

**Comment by pavolloffay** [comment]
I am wondering whether we should rename  `--collector.grpc-port` to something else like `--collector.grpc.host-port` or even scope that with service e.g. `batches` (though, it seems odd). The question is if we want to expose more services on the same port or use different grpc ports for different services. The first comment also includes 

> Combine gRPC and HTTP handlers on the same port


cc @yurishkuro @jpkrohling

**Comment by yurishkuro** [comment]
why do we need "host"?

**Comment by pavolloffay** [comment]
my bad there should not be host. The most important is scoping it under `.grpc.`

**Comment by yurishkuro** [comment]
For now I suggest keeping it simple without scoping. and serve multiple services on the same port.

**Comment by annanay25** [comment]
For the query-service, would it be better to multiplex the GRPC and HTTP handlers on the same port as compared to hosting two endpoints?

**Comment by yurishkuro** [comment]
yes, it is mentioned in the implementation plan

**Comment by annanay25** [comment]
Had a discussion with @pavolloffay and he mentioned it would be better from a security perspective to have different ports for the two endpoints. Would make sense if we would want to expose the GRPC port to public traffic and keep tchannel private.

**Comment by yurishkuro** [comment]
Why is it better for security? I can see that arguments for endpoints of different functionality, such as query vs. submitting spans, but not so much for the same functionality using different formats.

Anyway, we can certainly start with different ports. My preference is in the future to have the flexibility of the main() function recognizing that the same port is given for different protocols and automatically share the single server.

**Comment by annanay25** [comment]
I meant that if the HTTP endpoint of the query-service does not support SSL+authorisation, we would not be able to expose the multiplexed port publicly (unless of course we run it behind a reverse-proxy). Whereas GRPC would have HTTPS support. :)

**Comment by yurishkuro** [comment]
Very old issue, closing
